### ISSUE TITLE
Photon Gun Sample Changes Behavior after ~500,000 events?

### ISSUE THREAD

**Issue by jonzarling** [JeffersonLab/HDGeant4#94]
Hi all,

I'm trying to use geant4 to simulate a relatively large number of events within a single instance of calling hdgeant4 and seeing some odd behavior. More specifically, I am generating a sample of 1,000,000 photon gun events at theta=6 degrees and E_gamma=250 MeV.

Looking at the first 500,000 events, the distribution of primary photon conversion as a function of z makes sense (you can see interactions at target, CDC endplate, and FDC packages among other things)

![image](https://user-images.githubusercontent.com/17165552/53651077-ed03b880-3c13-11e9-8ad4-405f63a28000.png)

Looking at the last 500,000 events, the primary photon hardly ever escapes the target, so it would seem

![image](https://user-images.githubusercontent.com/17165552/53650688-16701480-3c13-11e9-9c1a-d8a16758463e.png)

I would be appreciative to anyone who could reproduce this effect. I've attached a control.in file and a python script (just remove .txt extension) to make the plots here. Thanks to Richard for the hddm_s parsing script. You should be able to just call hdgeant4 with this control.in file to generate photon gun MC yourself. The python script can then run over the output (pre-smeared) hddm file

[hddm_gun_parser.py.txt](https://github.com/JeffersonLab/HDGeant4/files/2920228/hddm_gun_parser.py.txt)
[control.in.txt](https://github.com/JeffersonLab/HDGeant4/files/2920229/control.in.txt)

**Comment by nsjarvis** [comment]
Hi Jon,

The diagrams in the email are the same.

I thought I'd try to run this partly selfishly as I need to bring my g4 up
to date.  Previously I have only run it with bggen.

What I have in my notes is that hdgeant4 will look for the control.in file,
but first it needs another input file as command-line arg which looks like

/run/beamOn 1000000
exit
EOI

Richard's wiki page says that "/run/beamOn 100" generates 100 minimum-bias
events using the built-in coherent bremsstrahlung generator as a source of
the simulated particles.

Does that look familiar?  What should that first line be for the particle
gun?


Best,
Naomi.

**Comment by jonzarling** [comment]
Hi Naomi,

Thanks for taking a look too. Sorry about the identical figures in the original thread, they should be updated on the github page. I supposed I was a little terse in explaining things here. That is correct, I also have a file in my running directory following MCWrapper's notation called run.mac with the following two lines inside to make 1,000,000 events

> /run/beamOn 1000000
> exit 

Then when running hdgeant4, the command takes the form:
>hdgeant4 -t4 run.mac

The control.in and run.mac files should be the only things (apart from environment setup) required for running hdgeant4.

Also if running the python script, then the following must be added to one's PYTHONPATH to be able to import hddm_s:
> setenv PYTHONPATH ${PYTHONPATH}:$HALLD_RECON_HOME/$BMS_OSNAME/python2

maybe that's already taken care of in others' environment setup, but it wasn't in mine. I think that's everything now.

**Comment by jonzarling** [comment]
Don't know if the "EOI" is important or not. Hopefully this is the correct usage for a particle gun where no beam photon is actually necessary.

**Comment by nsjarvis** [comment]
Great, that's what I have running.

If I have trouble with the python script I'll move the hddm file to
somewhere where you can get at it.  That Pythonpath add is new to me,
thanks.

Naomi.

**Comment by nsjarvis** [comment]
I cannot get the python script to work, it is complaining about lack of
ROOT.  I can try to fix that next week.

However the geant4 looks like it completed successfully.  I transferred the
hddm file to

/work/halld/njarvis/particle_gun.hddm

Please let me know what you find.

Best,
Naomi.

**Comment by jonzarling** [comment]
Ah, ROOT also requires an additional add-on to PYTHONPATH, guess that's not added by default either? Doing 
>setenv PYTHONPATH $ROOTSYS/lib:$PYTHONPATH

should add pyROOT and allow things to import.

I copied over the file you

**Comment by jonzarling** [comment]
I copied the file you generated. I see the same thing happen in your file, except that the flip in behavior seems to happen earlier than around event 500,000. More like event 125,000.

A plot of the first 125,000 events:
![image](https://user-images.githubusercontent.com/17165552/53668350-a462f400-3c41-11e9-8601-175e7de2c7d9.png)

A plot of the first 500,000 events (compare to plot at top of the thread)
![image](https://user-images.githubusercontent.com/17165552/53668323-901ef700-3c41-11e9-860b-4b6d3b2c0e0e.png)

A plot of the last 500,000 events
![image](https://user-images.githubusercontent.com/17165552/53668337-9a40f580-3c41-11e9-8650-bb59d2d3607a.png)

**Comment by rjones30** [comment]
Hello Noami and all,

The command run/beamOn 100 is what you use to launch the simulation in
hdgeant4, but everything else about the run is specified in the control.in
file.

Just read the comments in control.in and you should see how the various
options for the event generator get selected, between (a) and external
file, (b) the internal beam generator, and (c) the particle gun. I
understand that Thomas' mcwrapper has another scheme for doing similar
things, but I do not use mcwrapper myself. Knowing Thomas he probably has
written nice documentation on how to use mcwrapper.

-Richard

**Comment by nsjarvis** [comment]
This is weird.  Does the particle gun dry up?!  Should we run this at jlab
to make sure that it's not something odd with the off-site builds.

Naomi.

**Comment by nsjarvis** [comment]
Ah, thank you.  It too long since I have looked at any of this.  Jon's
question was a good excuse to get my code running at CMU, I hadn't used it
since the sim/recon split.

Best,
Naomi.

**Comment by rjones30** [comment]
Jon and Naomi, I thought surely this was a false alarm. I was wrong! 

I have reproduced the problem and am working on understanding it now. The symptom is that something happens after a large number of events have been simulated, that causes some process to start triggering a StopAndKill every time the primary particle exits the target, which it does through the sides or through the end depending on where the primary vertex is. The rogue process identifies itself as "compt" but I don't think it is legit Compton scattering. Anyway this is what you see in the z distribution you are plotting.

Don't spend any more time on this, I will publish a fix as soon as I get to the bottom of it.

-Richard

**Comment by rjones30** [comment]
Hello Jon and all,

What you uncovered turned out to be a nasty tracking bug that was abandoning the tracking of neutral particles under certain special circumstances. Although those circumstances were relatively rare, only once per 10k events or so, if they happened to occur on the last track in an event, then the tracking of all subsequent events would fail because there are only neutrals in your primary tracks in the special case you were studying: the photon gun generating gammas. I have submitted a new pull request that fixes this bug, and I ran a single run over 1M events to verify it works. Here is the plot that I get. Prior to this, I got the same results you did after event ~340,000. Now it runs to 1M events no problem.

![image](https://user-images.githubusercontent.com/7832920/53695391-3de8ed80-3d89-11e9-9e69-2217f678fdd3.png)

Please execute the pending pull request and verify that it works for you before closing this issue. I will follow up with a detailed description of the diagnosis and fix, in case you are interested.
-Richard

**Comment by rjones30** [comment]
Here is a detailed description of the bug, and the fix that eliminates it. You have to know something about the classes that are responsible for tracking in G4 to fully appreciate what is going on, but here it is in a nutshell. At the start of a new track, or the resumption of a suspended track, G4CoupledTransportation is supposed to clear the internal state of the G4PropagatorInField. However, this statement is bypassed in the case of the GlueX geometry because it is conditioned by if (fGlobalFieldExists) which is false for our geometry. GlueX has multiple magnetic fields (solenoid, pair spectrometer, sweeps) which are handled separately rather than as a single global field map. Because of this, the tracking of neutrals would fail in the following scenario.
           1) A charged particle stuck in some corner of a calorimeter block  triggers the infinite
              loop catching condition (fTrackIsLooping) in G4PropagatorInField,
              resulting in StopAndKill being generated for that track.
           2) The next particle is dequeued for tracking, and it happens
              to be neutral, so G4PropagatorInField::ComputeStep is not
              called for this track, and so the fTrackIsLooping remains set.
           3) G4PropagatorInField::ClearPropagatorState is normalled called
              at the start of each track, but only if fGlobalFieldExists
              is true. This fGlobalFieldExists is not true for the GlueX
              geometry because we have multiple field maps, eg. Solenoid,
              pair spectrometer, sweep magnets, etc. As a result, in our
              geometry the fTrackIsLooping condition persists until the
              next charged particle shows up to be tracked.
           4) The AlongStepGetPhysicalInteractionLength method of the
              main track following class G4CoupledTransportation does a
              StopAndKill on any track whenever the G4PropagatorInField reports
              that fTrackIsLooping==true, regardless of whether it is
              charged or not.
The upshot of all this is that all neutrals that come up after a charged particle is killed off for looping too long get dropped until the next charged particle shows up for tracking. This is normally not too bad because the looping condition hardly ever happens (once per few 10k events or so) and even when it does, if there are charged particles still on the stack to be tracked, the condition will be reset as soon as G4PropagateInField::ComputeStep is invoked. ***BUT*** if that looping particle happens to be the last one in an event, and the next event has only neutral primaries in it (eg. gammas from the particle gun) then all of the tracking will do StopAndKill on the first step for ALL REMAINING EVENTS in the simulation!!!

The fix was simple: remove the if (fGlobalFieldExists condition around the ClearPropagatorState at the start of each new track. When starting a new track, when would you ever want the state of the previous track to be remembered?  I also apply the same fix to G4Transportation where the same pathology also appears. Although it does not seem to affect our hdgeant4 sims at present, it may in the future.

**Comment by jonzarling** [comment]
Hi Richard,

I built the master hdgeant4 with your recent updates. Looks like everything is working again. No sign of the issue, and file sizes are much larger now (as expected with photons that escape the target!).

![image](https://user-images.githubusercontent.com/17165552/53769709-bdbaa900-3eaa-11e9-8246-e3b7d5593d94.png)

Thanks,
Jon
